From 737b7c7e99cfb3f9c065ddc8e4c23af67811f088 Mon Sep 17 00:00:00 2001
From: Hans Rosenfeld <hans.rosenfeld@joyent.com>
Date: Wed, 1 Aug 2018 15:25:26 +0000
Subject: [PATCH] OS-6966 SmartOS hangs in boot on bhyve due to recursive mutex
 in pci_virtio_block.c Reviewed by: Mike Gerdts <mike.gerdts@joyent.com>
 Reviewed by: John Levon <john.levon@joyent.com> Reviewed by: Patrick Mooney
 <patrick.mooney@joyent.com> Approved by: John Levon <john.levon@joyent.com>

---
 usr/src/cmd/bhyve/virtio.h | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/usr/src/cmd/bhyve/virtio.h b/usr/src/cmd/bhyve/virtio.h
index cd28f55738..f59d823448 100644
--- a/usr/src/cmd/bhyve/virtio.h
+++ b/usr/src/cmd/bhyve/virtio.h
@@ -31,6 +31,8 @@
 #ifndef	_VIRTIO_H_
 #define	_VIRTIO_H_
 
+#include <pthread_np.h>
+
 /*
  * These are derived from several virtio specifications.
  *
@@ -438,11 +440,25 @@ vq_interrupt(struct virtio_softc *vs, struct vqueue_info *vq)
 	if (pci_msix_enabled(vs->vs_pi))
 		pci_generate_msix(vs->vs_pi, vq->vq_msix_idx);
 	else {
+#ifndef __FreeBSD__
+		boolean_t unlock = B_FALSE;
+
+		if (vs->vs_mtx && !pthread_mutex_isowned_np(vs->vs_mtx)) {
+			unlock = B_TRUE;
+			pthread_mutex_lock(vs->vs_mtx);
+		}
+#else
 		VS_LOCK(vs);
+#endif
 		vs->vs_isr |= VTCFG_ISR_QUEUES;
 		pci_generate_msi(vs->vs_pi, 0);
 		pci_lintr_assert(vs->vs_pi);
+#ifndef __FreeBSD__
+		if (unlock)
+			pthread_mutex_unlock(vs->vs_mtx);
+#else
 		VS_UNLOCK(vs);
+#endif
 	}
 }
 
-- 
2.21.0

